File: LINKED LIST/Detect Loop/DetectLoop.cpp
#include <unordered_map>

class Node {
public:
    int data;
    Node *next;

    Node() {
        this->data = 0;
        next = nullptr;
    }

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }

    Node(int data, Node* next) {
        this->data = data;
        this->next = next;
    }
};

bool detectCycle(Node* head) {
    std::unordered_map<Node*, char> visited;
    Node* slow = head;
    Node* fast = head;
    while (fast != nullptr) {
        slow = slow->next;
        fast = fast->next;
        if (fast != nullptr) {
            fast = fast->next;
        }
        if (slow == fast && slow != nullptr) {
            return true;
        }
    }
    return false;
}

File: LINKED LIST/Detect Loop/DetectLoop.java
import java.util.HashMap;

class Node {
    public int data;
    public Node next;

    Node() {
        this.data = 0;
        this.next = null;
    }

    Node(int data) {
        this.data = data;
        this.next = null;
    }

    Node(int data, Node next) {
        this.data = data;
        this.next = next;
    }
}

public class DetectLoop {
    // This method is called tortoise and hare method or Floyd's cycle detection algorithm
    // Another approach is hashing
    public static boolean detectCycle(Node head) {
        HashMap<Node, Character> visited = new HashMap<>();
        Node slow = head;
        Node fast = head;
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
            if (fast != null) {
                fast = fast.next;
            }
            if (slow == fast && slow != null) {
                return true;
            }
        }
        return false;
    }
}
